### DISCUSSION TITLE
Support for async-graphql

### DISCUSSION THREAD

**Discussion by tqwewe** [SeaQL/sea-orm#87] [Ideas]
With SeaORM made public now, it opens a path to be used in Rust GraphQL servers with a top library being async-graphql.
Having seamless integration with async-graphql would be amazing.

I managed to setup basic support for it by adding the following to `sea-orm-macros/src/derives/active_model.rs`:
```rust
#[derive(Clone, Debug, PartialEq)]
#[cfg_attr(feature = "graphql", derive(InputObject))]
pub struct ActiveModel {
    #(pub #field: sea_orm::ActiveValue<#ty>),*
}
```
Which makes an `ActiveModel` derive [`InputObject`](https://docs.rs/async-graphql/2.9.10/async_graphql/derive.InputObject.html).

I had to additionally add the following implementations:
`impl<T> InputType for ActiveValue<T>`
`impl<T> InputType for ActiveValue<Option<T>>`
`impl<T> Type for ActiveValue<T>`

Where the parse function for `ActiveValue<T>` is:
```rust
default fn parse(value: Option<Value>) -> InputValueResult<Self> {
    match value {
        None | Some(Value::Null) => Ok(ActiveValue::unset()), // Column does not support setting to null, so just unset
        Some(value) => Ok(ActiveValue::set(
            T::parse(Some(value)).map_err(InputValueError::propagate)?,
        )),
    }
}
```
And for `ActiveValue<Option<T>>` is:
```rust
fn parse(value: Option<Value>) -> InputValueResult<Self> {
    match value {
        None => Ok(ActiveValue::unset()),
        Some(Value::Null) => Ok(ActiveValue::set(None)), // Value is 'null', we can update the DB row to null
        Some(value) => Ok(ActiveValue::set(Some(
            T::parse(Some(value)).map_err(InputValueError::propagate)?,
        ))),
    }
}
```
Unfortunately with this approach, I required specialization to be enabled, so had to switch rustup to nightly for this to work.

This was all implemented under a feature flag `"graphql"`.

Once implemented, graphql mutations can accept an `ActiveModel` of an entity for updating a subset of columns, and setting some columns to `null`.
```rust
async fn update_cake(
    &self,
    ctx: &Context<'_>,
    cake: cakes::ActiveModel,
) -> Result<bool> {
    let db = ctx.data::<DatabaseConnection>()?;

    Cake::update(cake).exec(db).await;

    Ok(true)
}
```

**Comment by tyt2y3** [comment]
Thank you for the information.
However I am not too sure about the 'default' keyword and the 'specialization' of Rust.
What are their uses?

**Reply by tqwewe** [reply]
I have a basic understanding... pretty much because `impl<T> InputType for ActiveValue<T>` is more generic than `impl<T> InputType for ActiveValue<Option<T>>`, we can label the parse function as `default` which makes it act like a default implementation... and we override the implementation for for specific types such as `ActiveValue<Option<T>>`.

https://rust-lang.github.io/rfcs/1210-impl-specialization.html

**Comment by RustLoverTheCoder** [comment]
Is it supported now?

**Reply by billy1624** [reply]
Hi @Toursslivers, not yet supported

**Reply by baoyachi** [reply]
> Is it supported now?

We look forward to your contribute. 😃

**Reply by tyt2y3** [reply]
It's certainly workable to use SeaORM to implement GraphQL resolvers.
Which @Acidic9 has been doing.
It's just not been included in our test suite and examples.
Ari, how's your experience so far?
It'd be great if you can share a basic integration example!

**Reply by tqwewe** [reply]
I have taken a break from it for the past few days on my personal project, but what I ended up doing was this:

With PR https://github.com/SeaQL/sea-orm/pull/240, I create new structs for inputs:
```rust
// Input for Graphql queries and mutations (with InputObject)
#[derive(InputObject, IntoActiveModel)]
pub struct NewUser {
    pub email: String,
    pub password: String,
}
```

And for outputs, I just derive `SimpleObject` on my main model:
```rust
// Output for Graphql queries and mutations (with SimpleObject)
#[derive(DeriveEntityModel, DeriveActiveModelBehavior, SimpleObject, Clone, Debug)]
#[sea_orm(table_name = "user")]
#[graphql(concrete(name = "User", params()))]
pub struct Model {
    #[sea_orm(primary_key, auto_increment = false, unique)]
    pub id: Uuid,
    pub created_at: DateTimeWithTimeZone,
    pub updated_at: DateTimeWithTimeZone,
    pub email: String,
    pub password: String,
}
```

This way, sea orm does not need to do anything for graphql support, and it keeps the two separated. I think adding in-build graphql support might actually lead to SeaORM becoming bloated.

**Reply by hyperliskdev** [reply]
Hi, I know this thread is old, but I am currently trying to use the DateTime type with SimpleObject but it seems the version of datetime that SeaORM uses does not support the SimpleObject derive. 

Does SeaORM have another supported version of DateTime that I can use or should I implement my own?
`
async_graphql_derive
pub macro derive_simple_object
the trait bound sea_orm::prelude::DateTime: async_graphql::Type is not satisfied
the trait async_graphql::Type is not implemented for sea_orm::prelude::DateTime
`

**Reply by shanesveller** [reply]
@hyperliskdev I looked at a stagnant project of mine using `async-graphql` and `sea-orm` and in November 2021-ish I was able to use `Date` types from `chrono` (re-exported by `sea_orm`) via a Cargo.toml entry to activate its serde crate feature:

```toml
chrono = { version = "...", features = ["serde"] }
```

See if that helps?

**Reply by billy1624** [reply]
Thanks! @shanesveller

**Comment by RustLoverTheCoder** [comment]
Thanks, I have used it in my project
